pkg/controller/mizar/mizar-node-controller.go

```diff
@@ -0,0 +1,281 @@
+			} else {
+				eventType, err := c.determineEventType(oldKey, newKey)
+				if err != nil {
+					klog.Errorf(c.logMessage("Unexpected string in queue; discarding: ", newKey))
+				} else {
+					c.Enqueue(newKey, eventType)
+				}
+				klog.Infof(c.logMessage("Update Node: ", newKey))
+			}
+		},
+		DeleteFunc: func(object interface{}) {
+			resource := object.(*v1.Node)
+			key := c.genKey(resource)
+			c.Enqueue(key, EventType_Delete)
+			klog.Infof(c.logMessage("Delete Node: ", key))
+		},
+	})
+	return c, nil
+}
+
+// Run starts an asynchronous loop that detects events of cluster nodes.
+func (c *MizarNodeController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+	klog.Infof(c.logInfoMessage("Starting node controller"))
+	klog.Infof(c.logInfoMessage("Waiting cache to be synced"))
+	if ok := cache.WaitForCacheSync(stopCh, c.informerSynced); !ok {
+		klog.Fatalf(c.logInfoMessage("Timeout expired during waiting for caches to sync."))
+	}
+	klog.Infof(c.logInfoMessage("Starting workers..."))
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+	<-stopCh
+	klog.Infof(c.logInfoMessage("Shutting down node controller"))
+}
+
+// Enqueue puts key of the node object in the work queue
+// EventType: Create=0, Update=1, Delete=2, Resume=3
+func (c *MizarNodeController) Enqueue(key string, eventType EventType) {
+	c.queue.Add(KeyWithEventType{Key: key, EventType: eventType})
+}
+
+// Dequeue an item and process it
+func (c *MizarNodeController) runWorker() {
+	for {
+		item, queueIsEmpty := c.queue.Get()
+		if queueIsEmpty {
+			break
+		}
+		c.process(item)
+	}
+}
+
+// Parsing a item key and call gRPC request
+func (c *MizarNodeController) process(item interface{}) {
+	defer c.queue.Done(item)
+	keyWithEventType, ok := item.(KeyWithEventType)
+	if !ok {
+		klog.Errorf(c.logMessage("Unexpected item in queue: ", keyWithEventType))
+		c.queue.Forget(item)
+		return
+	}
+	//command, key, err := c.parseCommandKey(commandKey)
+	key := keyWithEventType.Key
+	eventType := keyWithEventType.EventType
+	nodeKind, nodeTenant, nodeName, nodeStatusType, nodeStatus, nodeAddressType, nodeAddress, err := c.parseKey(key)
+	if err != nil || nodeKind != NodeKind {
+		klog.Errorf(c.logMessage("Unexpected string in queue; discarding: ", key))
+		c.queue.Forget(item)
+		return
+	}
+	klog.Infof(c.logInfoMessage("Processing a node: ")+"%s/%s/%s/%s/%s/%s/%s", nodeKind, nodeTenant, nodeName, nodeStatusType, nodeStatus, nodeAddressType, nodeAddress)
+	c.gRPCRequest(eventType, nodeName, nodeAddress)
+	c.queue.Forget(item)
+}
+
+// Generate a node key
+func (c *MizarNodeController) genKey(resource *v1.Node) string {
+	nodeKind := resource.GetObjectKind().GroupVersionKind().Kind
+	if nodeKind == "" {
+		nodeKind = NodeKind
+	}
+	nodeTenant := resource.GetTenant()
+	nodeName := resource.GetName()
+	if nodeName == "" {
+		klog.Fatalln(c.logMessage("Node name should not be null", nodeName))
```

Comment:
We cannot use fatalln here. We need to make sure node controller don't crash the kube-controller-manager process.